### PR TITLE
CPU offload toggles feature

### PR DESCRIPTION
**Pull request by ShotaKaji5207**
Testing in progress. 

Adds `deepspeed_cpu_adam` option to enable DeepSpeedCPUAdam optimizer.

Addresses #1031

### CODE REVIEW

**Comment by hamishivi** [comment]
Hi @ShotaKaji5207 are you still working on this? No worries if not, just cleaning up open PRs :)

**Comment by ShotaKaji5207** [comment]
Hi @hamishivi , thanks!

I'm not working on this currently. grpo_fast.sh runs through in my environment without CPU offloading now.

**Comment by ShotaKaji5207** [comment]
Hi @hamishivi 
There is an update.
I'm running it again on the latest repo and doing ablation experiments on resource usage.

**Comment by ShotaKaji5207** [comment]
## Summary

I ran an ablation study to evaluate the effect of DeepSpeed CPU offload toggles.

- In my setup, **`deepspeed_cpu_adam` + `deepspeed_offload_optimizer`** provided the best overall trade-off (lower peak GPU memory and faster wall-clock time), with consistent trends under both ZeRO stage 2 and ZeRO stage 3.
- The option description in `grpo_utils.py` for `deepspeed_cpu_adam` notes that it requires `--deepspeed_offload_optimizer`.

**Setup:** `grpo_fast.sh` + `Qwen/Qwen2.5-0.5B-Instruct` (single GPU, ZeRO 2/3); varied toggles: `deepspeed_cpu_adam`, `deepspeed_offload_optimizer`, `deepspeed_offload_param`.

---

## ZeRO Stage 2: DeepSpeed offload toggles ablation

### Completed runs

| Setting | deepspeed_cpu_adam | deepspeed_offload_optimizer | deepspeed_offload_param | Peak GPU mem (MiB) | Peak CPU (%) | Wall-clock time | Notes |
|---|---:|---:|---:|---:|---:|---:|---|
| Baseline | false | false | false | 16000 | 11.51 | 24m 43s | — |
| DeepSpeedCPUAdam + optimizer offload | true | true | false | 11601 | 13.69 | 18m 10s | — |
| Parameter offload only | false | false | true | 16034 | 11.52 | 31m 24s | — |
| Full offload | true | true | true | 11657 | 14.31 | 20m 33s | — |

### Not runnable / blocked configs

- **DeepSpeedCPUAdam only** (true, false, false): **Not runnable**. `deepspeed_cpu_adam` requires `--deepspeed_offload_optimizer`.
- **DeepSpeedCPUAdam + parameter offload** (true, false, true): **Not runnable**. Same as above (confirmed in experiments).
- **Optimizer offload only** (false, true, false): **Blocked by default (guard)** for ZeRO-Offload + client optimizer. Can be run by setting `zero_force_ds_cpu_optimizer: false`.  
  - Not included in the tables: an older run recorded peak GPU mem 6581 MiB and peak CPU 11.63%, but these values are **reference-only** due to unknown conditions.
- **Optimizer offload + parameter offload** (false, true, true): **Blocked by default (guard)**. Can be run by setting `zero_force_ds_cpu_optimizer: false`.

---

## ZeRO Stage 3: DeepSpeed offload toggles ablation

| Setting | deepspeed_cpu_adam | deepspeed_offload_optimizer | deepspeed_offload_param | Peak GPU mem (MiB) | Peak CPU (%) | Wall-clock time | Notes |
|---|---:|---:|---:|---:|---:|---:|---|
| Baseline | false | false | false | 15998 | 11.78 | 24m 57s | — |
| Parameter offload only | false | false | true | 16023 | 11.22 | 30m 06s | — |
| DeepSpeedCPUAdam + optimizer offload | true | true | false | 14513 | 14.78 | 18m 47s | — |
| Full offload | true | true | true | — | — | — | **CPU OOM** (node RAM 31 GB was insufficient) |

---

## Notes (footnotes)

- **"Blocked by default (guard)"** refers to the DeepSpeed/Transformers integration stopping the combination of ZeRO-Offload and a **client optimizer** (e.g., `torch.optim.AdamW`). Setting `zero_force_ds_cpu_optimizer: false` can override this behavior.
- Measurements labeled **reference-only** indicate potential mismatch between recorded toggles and the actual runtime configuration (unknown conditions).

---

<details>
<summary>Full launch script (reference)</summary>

```bash
#!/bin/bash
export VLLM_ALLOW_INSECURE_SERIALIZATION=1
export VLLM_DISABLE_COMPILE_CACHE=1
export VLLM_USE_V1=1

# Optional: W&B (kept offline for this run)
export WANDB_MODE=offline
# export WANDB_ENTITY=...
# export WANDB_PROJECT=...

# Unset VIRTUAL_ENV to prevent Ray workers from creating new virtual environments
unset VIRTUAL_ENV

python open_instruct/grpo_fast.py \
    --dataset_mixer_list ai2-adapt-dev/rlvr_gsm8k_zs 64 \
    --dataset_mixer_list_splits train \
    --max_prompt_token_length 256 \
    --response_length 256 \
    --pack_length 512 \
    --per_device_train_batch_size 1 \
    --num_unique_prompts_rollout 2 \
    --num_samples_per_prompt_rollout 4 \
    --model_name_or_path Qwen/Qwen2.5-0.5B-Instruct \
    --load_ref_policy false \
    --stop_strings "</answer>" \
    --apply_verifiable_reward true \
    --temperature 0.7 \
    --ground_truths_key ground_truth \
    --chat_template_name r1_simple_chat_postpend_think \
    --learning_rate 3e-7 \
    --total_episodes 100 \
    --deepspeed_stage 3 \
    --num_epochs 1 \
    --num_learners_per_node 1 \
    --vllm_tensor_parallel_size 1 \
    --beta 0.00 \
    --seed 3 \
    --local_eval_every -1 \
    --vllm_sync_backend gloo \
    --vllm_gpu_memory_utilization 0.3 \
    --vllm_enforce_eager \
    --gradient_checkpointing \
    --gather_whole_model false \
    --deepspeed_offload_optimizer true \
    --deepspeed_cpu_adam true \
    --deepspeed_offload_param true \
    --single_gpu_mode \
    --push_to_hub false \
    --system_prompt_override_file scripts/train/debug/cute_debug_system_prompt.txt \
    --active_sampling --async_steps 4 \
    --with_tracking \
    --verbose

**Comment by hamishivi** [comment]
closing since not much activity recently. If you really want to add cpu offloading please reopen!

**Comment by ShotaKaji5207** [comment]
Thanks for the cleanup and for maintaining this repository!
